iceberg/src/main/scala/com/nvidia/spark/rapids/iceberg/parquet/GpuParquetReaderPostProcessor.scala

```diff
@@ -0,0 +1,177 @@
+      "iceberg post processing")) { batch =>
+      withResource(new ColumnarBatchHandler(this, batch)) { handler =>
+        TypeUtil.visit(expectedSchema, handler).left.get
+      }
+    }
+  }
+}
+
+
+private class ColumnarBatchHandler(private val processor: GpuParquetReaderPostProcessor,
+    private val batch: LazySpillableColumnarBatch
+) extends TypeUtil.SchemaVisitor[HandlerResult] with AutoCloseable {
+  batch.allowSpilling()
+  private var currentField: NestedField = _
+  // This is used to hold the column vectors that are created during the travel of the batch, so
+  // that if exception happens, we can close them all at once.
+  private val vectorBuffer: ArrayBuffer[GpuColumnVector] = new ArrayBuffer[GpuColumnVector](
+    batch.numCols)
+
+  override def schema(schema: Schema, structResult: HandlerResult): HandlerResult  = structResult
+
+  override def struct(struct: Types.StructType, fieldResults: util.List[HandlerResult])
+  : HandlerResult = {
+    val columns = new Array[ColumnVector](fieldResults.size)
+    for (i <- 0 until fieldResults.size) {
+      columns(i) = fieldResults.get(i).right.get
+    }
+    // Ownership has transferred to columnar batch, so it should be cleared
+    vectorBuffer.clear()
+    Left(new ColumnarBatch(columns, batch.numRows))
+  }
+
+  override def field(field: Types.NestedField, fieldResult: HandlerResult): HandlerResult = {
+    if (!field.`type`.isPrimitiveType) {
+      throw new UnsupportedOperationException("Unsupported type" +
+        " for iceberg scan: " + field.`type`)
+    }
+    fieldResult
+  }
+
+
+  override def beforeField(field: Types.NestedField): Unit = {
+    currentField = field
+  }
+
+  override def primitive(fieldType: Type.PrimitiveType): HandlerResult = {
+    val gpuColVector = doProcessPrimitive(fieldType)
+    vectorBuffer += gpuColVector
+    Right(gpuColVector)
+  }
+
+  private def doProcessPrimitive(fieldType: Type.PrimitiveType): GpuColumnVector = {
+    val curFieldId = currentField.fieldId
+    val sparkType = SparkSchemaUtil.convert(fieldType)
+    // need to check for key presence since associated value could be null
+    if (processor.idToConstant.containsKey(curFieldId)) {
+      withResource(GpuScalar.from(processor.idToConstant.get(curFieldId), sparkType)) { scalar =>
+        return GpuColumnVector.from(scalar, batch.numRows, sparkType)
+      }
+    }
+    if (curFieldId == MetadataColumns.ROW_POSITION.fieldId) {
+      throw new UnsupportedOperationException("ROW_POSITION meta column is not supported yet")
+    }
+    if (curFieldId == MetadataColumns.IS_DELETED.fieldId) {
+      throw new UnsupportedOperationException("IS_DELETED meta column is not supported yet")
+    }
+
+    for (i <- 0 until processor.fileReadSchema.getFieldCount) {
+      val t = processor.fileReadSchema.getType(i)
+      if (t.getId != null && t.getId.intValue == curFieldId) {
+        return doUpCastIfNeeded(batch.getBatch.column(i).asInstanceOf[GpuColumnVector], fieldType)
```

Comment:
Does this leak the batch? I don't remember.

Reply:
if `batch` is lazy, no. But otherwise, I think so.